client/file-browser/browse-files.jsx

```diff
@@ -0,0 +1,526 @@
+        )
+      } else if (entry.type === 'folder') {
+        return (
+          <FolderEntry
+            style={style}
+            folder={entry}
+            onClick={this.onFolderClick}
+            isFocused={isFocused}
+            key={entry.path}
+          />
+        )
+      } else if (entry.type === 'file') {
+        return (
+          <FileEntry
+            style={style}
+            file={entry}
+            onClick={this.onFileClick}
+            icon={fileTypes[entry.extension].icon}
+            isFocused={isFocused}
+            key={entry.path}
+          />
+        )
+      } else {
+        throw new Error('Invalid entry type: ' + entry.type)
+      }
+    }
+
+    const width = this.contentRef.current.getClientWidth()
+    const height = this.contentRef.current.getClientHeight()
+    return (
+      <>
+        <KeyListener onKeyDown={this.onKeyDown} />
+        <VirtualizedList
+          ref={this.listRef}
+          width={width}
+          height={height - VERT_PADDING * 2}
+          rowCount={entries.size}
+          rowHeight={ENTRY_HEIGHT}
+          rowRenderer={_rowRenderer}
+          className={styles.fileList}
+          style={{ overflowX: false, overflowY: false }}
+          containerStyle={{ marginTop: VERT_PADDING, marginBottom: VERT_PADDING }}
+        />
+      </>
+    )
+  }
+
+  render() {
+    const { rootFolderName, title, root, error } = this.props
+    const { path } = this.props.fileBrowser[this.props.browseId]
+    const displayedPath = `${rootFolderName}\\${pathApi.relative(root, path)}`
+    return (
+      <div ref={this.browserRef} tabIndex='-1' className={styles.root}>
+        <div className={styles.topBar}>
+          <div className={styles.title}>
+            <h3 className={styles.contentTitle}>{title}</h3>
+          </div>
+          <div className={styles.breadcrumbsAndActions}>
+            <PathBreadcrumbs
+              className={styles.path}
+              path={displayedPath}
+              onNavigate={this.onBreadcrumbNavigate}
+            />
+            <IconButton icon={<Refresh />} onClick={this.onRefreshClick} title={'Refresh'} />
+          </div>
+        </div>
+        {error ? <div className={styles.externalError}>{error}</div> : null}
+        <ScrollableContent
+          ref={this.contentRef}
+          className={styles.filesScrollable}
+          viewClassName={styles.filesScrollableView}
+          onScroll={this.onScroll}>
+          {this.renderFiles()}
+        </ScrollableContent>
+      </div>
+    )
+  }
+
+  onBreadcrumbNavigate = path => {
+    const { root } = this.props
+    const pathWithoutRoot = path.slice(this.props.rootFolderName.length + 1)
+    this.props.dispatch(changePath(this.props.browseId, pathApi.join(root, pathWithoutRoot)))
+  }
+
+  onUpLevelClick = () => {
+    const { path } = this.props.fileBrowser[this.props.browseId]
+    const prevPath = path.lastIndexOf('\\') !== -1 ? path.slice(0, path.lastIndexOf('\\')) : ''
```

Comment:
do we need to worry about different types of slashes here? would it be better to use the `path` module to do these kind of things?

Reply:
Well, we only support Windows for now :P
But yeah, we use the `path` module to initially [set the path](https://github.com/ShieldBattery/ShieldBattery/blob/f699775c38c5a903bff61db484f3d2677043945c/client/replays/get-files.js#L11), so the path separator will be platform-specific. I'll update the handling of separators in this file to use the `path` module then, to be more future proof.